src/realm/sync/network/default_socket.hpp

```diff
@@ -43,49 +45,46 @@ class DefaultSocketProvider : public SyncSocketProvider {
         network::DeadlineTimer m_timer;
     };
 
-    DefaultSocketProvider(const std::shared_ptr<util::Logger>& logger, const std::string user_agent)
-        : m_logger_ptr{logger}
-        , m_service{std::make_shared<network::Service>()}
-        , m_random{}
-        , m_user_agent{user_agent}
-    {
-        REALM_ASSERT(m_logger_ptr);                     // Make sure the logger is valid
-        REALM_ASSERT(m_service);                        // Make sure the service is valid
-        util::seed_prng_nondeterministically(m_random); // Throws
-        start_keep_running_timer();
-    }
+    DefaultSocketProvider(const std::shared_ptr<util::Logger>& logger, const std::string user_agent);
 
     // Don't allow move or copy constructor
     DefaultSocketProvider(DefaultSocketProvider&&) = delete;
 
-    // Temporary workaround until event loop is completely moved here
-    void run() override
-    {
-        m_service->run();
-    }
+    ~DefaultSocketProvider();
 
-    void stop() override
-    {
-        m_service->stop();
-    }
+    /// Temporary workaround until client shutdown has been updated in a separate PR - these functions
+    /// will be handled internally when this happens.
+    /// Stops the internal event loop (provided by network::Service)
+    void stop(bool wait_for_stop = false) override;
 
     std::unique_ptr<WebSocketInterface> connect(WebSocketObserver*, WebSocketEndpoint&&) override;
 
     void post(FunctionHandler&& handler) override
     {
-        REALM_ASSERT(m_service);
         // Don't post empty handlers onto the event loop
         if (!handler)
             return;
-        m_service->post(std::move(handler));
+        m_service.post(std::move(handler));
     }
 
     SyncTimer create_timer(std::chrono::milliseconds delay, FunctionHandler&& handler) override
     {
-        return std::unique_ptr<Timer>(new DefaultSocketProvider::Timer(*m_service, delay, std::move(handler)));
+        return std::unique_ptr<Timer>(new DefaultSocketProvider::Timer(m_service, delay, std::move(handler)));
     }
 
 private:
+    enum class State : int { Starting, Started, Running, Stopping, Stopped };
```

Comment:
there's no need to specify the underlying type of an enum unless you have specific storage requirements - like you need to know it's an unsigned 64-bit integer because you're going to interpret it as a bitfield. You can just do `enum class State {}`